_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+
+How are we going to train this model? Method-wise, we have a choice between variational inference (VI) and Hamiltonian Monte Carlo (HMC). We will see both. The second question is: Are we going to use graph mode or eager mode? As of today, for both VI and HMC, it is safest -- and fastest -- to run in graph mode, so this is the only technique we show. In a few weeks, or months, we should be able to prune a lot of `sess$run()`s from the code!
+
+Normally in posts, when presenting code we optimize for easy experimentation (meaning: line-by-line executability), not modularity. This time though, with an important number of evaluation statements involved, it's easiest to pack not just the fitting, but the smoothing and forecasting as well into a function (which you could still step through if you wanted). For VI, we'll have a `fit _with_vi` function that does it all. So when we now start explaining what it does, don't type in the code just yet -- it'll all reappear nicely packed into that function, for you to copy and execute as a whole.
+
+#### Fitting a time series with variational inference
+
+Fitting with VI pretty much looks like training traditionally used to look in graph-mode TensorFlow. You define a loss -- here it's done using [sts_build_factored_variational_loss()](https://rstudio.github.io/tfprobability/reference/sts_build_factored_variational_loss.html) --, an optimizer, and an operation for the optimizer to reduce that loss:
+
+```{r}
+optimizer <- tf$compat$v1$train$AdamOptimizer(0.1)
+
+# only train on the training set!    
+loss_and_dists <- ts_train %>% sts_build_factored_variational_loss(model = model)
+variational_loss <- loss_and_dists[[1]]
+
+train_op <- optimizer$minimize(variational_loss)
+```
+
+Note how the loss is defined on the training set only, not the complete series.
+
+Now to actually train the model, we create a session and run that operation:
+
+```{r}
+ with (tf$Session() %as% sess,  {
+      
+      sess$run(tf$compat$v1$global_variables_initializer())
+   
+      for (step in 1:n_iterations) {
+        res <- sess$run(train_op)
+        loss <- sess$run(variational_loss)
+        if (step %% 10 == 0)
+          cat("Loss: ", as.numeric(loss), "\n")
+      }
+ })
+```
+
+
+Given we have that session, let's make use of it and compute all the estimates we desire.
+Again, -- the following snippets will end up in the `fit_with_vi` function, so don't run them in isolation just yet.
+
+#### Obtaining forecasts
+
+The first thing we want for the model to give us are forecasts. In order to create them, it needs _samples from the posterior_. Luckily we already have the posterior distributions, returned from `sts_build_factored_variational_loss`, so let's sample from them and pass them to [sts_forecast](https://rstudio.github.io/tfprobability/reference/sts_forecast.html):
```

Comment:
> The first thing we want for the model to give us are forecasts.

Hmm, this one is tricky for subject-verb agreement; "thing" is singular, but "forecasts" is plural (so you're conjugating with _are_). It's your choice — I might avoid it altogether and use something like _First we want to get forecasts from the model._